hypothesis-python/src/hypothesis/internal/conjecture/data.py

```diff
@@ -200,9 +200,11 @@ def structural_coverage(label: int) -> StructuralCoverageTag:
 NASTY_FLOATS = list(map(float, NASTY_FLOATS))
 NASTY_FLOATS.extend([-x for x in NASTY_FLOATS])
 
-FLOAT_INIT_LOGIC_CACHE = LRUReusedCache(4096)
-
-POOLED_KWARGS_CACHE = LRUReusedCache(4096)
+# These caches, especially the kwargs cache, can be quite hot and so we prefer
+# LRUCache over LRUReusedCache for performance. We lose scan resistance, but
+# that's probably fine here.
+FLOAT_INIT_LOGIC_CACHE = LRUCache(4096)
+POOLED_KWARGS_CACHE = LRUCache(4096)
```

Comment:
I am wondering if these caches could be thread-local? to avoid races as resolved by #4013

Reply:
Thanks for pointing this out! Changed to threadlocal, though I haven't explicitly tested it with multithreading.